### PR TITLE
Fix NPE in ControlConnection.updateLocationInfo (JAVA-656)

### PR DESCRIPTION
**Pull request by tommystendahl**

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java [hunk 1]
```diff
@@ -568,7 +568,11 @@ private static void refreshNodeListAndTokenMap(Connection connection, Cluster.Ma
                 // We don't know that node, create the Host object but wait until we've set the known
                 // info before signaling the addition.
                 host = cluster.metadata.add(foundHosts.get(i));
-                isNew = true;
+                if (host == null)
+                    // There must have been a race condition and someone else added the node before us
+                    host = cluster.metadata.getHost(foundHosts.get(i));
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java]: This is not atomic with the last call to add. I know it's highly unlikely, but in theory host could be null again after this line.

We could split `Metadata.add` into two methods:
- `Metadata.newHost(InetSocketAddress)` that creates the Host object
- `Metadata.addIfAbsent(Host)` with putIfAbsent semantics (returns the existing Host if it already existed)

Then we could write this:

``` java
// We don't know that node, create the Host object but wait until we've set the known
// info before signaling the addition.
Host newHost = cluster.metadata.newHost(foundHosts.get(i));
Host existing = cluster.metadata.addIfAbsent(newHost);
if (existing == null) {
    host = newHost;
    isNew = true;
} else {
    host = existing;
    isNew = false;
}
```

Then we need to address the other places where `add` was used, but that's easy.

Comment (tommystendahl) [inline driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java]: Yes, your right. Doing it this way would close the last small hole and it should be easy enough to do. I have an updated patch up soon.

**Comment by tommystendahl** [comment]
I rewrote the `metadata.add(InetSocketAddress)` to use the `addIfAbsent` and `newHost` instead of removing it. I think this made the code in the other places that calls add cleaner, insted of doing `metadata.addIfAbsent(metadata.newHost(address))` they can continue doing `metadata.add(address)`.

File: changelog/README.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 ### 2.0.12 (in progress)
 
+- [bug] NPE in ControlConnection.updateLocationInfo (JAVA-656)
```
Comment (olim7t) [inline changelog/README.md]: Could you add your entry at the end of the list rather than the beginning?
It's just that it makes merges easier.

Comment (tommystendahl) [inline changelog/README.md]: No problem.

**Comment by tolbertam** [comment]
Looks good to me and seems straightforward :+1: on merging.  I think I should be able to reproduce the original problem by finding a reliable way to generate multiple add/remove/move events simultaneously causing a new host to be detected in multiple executor threads at a time.  Will develop a scenario around that as it seems useful to have.

**Comment by olim7t** [comment]
Cherry-picked as a9f9373 (I should have merged your PR first to avoid conflicts, sorry about that).

**Pull request closed without merging**